### PR TITLE
Read cache job 1

### PR DESCRIPTION
**Pull request by sethiay**
### Description
Add structure for File Download Job struct. Currently, I haven't added unimplemented methods for job.Start() and job.Retry as I am figuring out their final function signatures after today's call.

### Link to the issue in case of a bug fix.
NA

### Testing details
1. Manual - Ran unit tests.
2. Unit tests - NA
3. Integration tests - NA

### CODE REVIEW

File: internal/cache/file/downloader/job.go [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+	/////////////////////////
+	// Constant data
+	/////////////////////////
+
+	object               *gcs.MinObject
+	bucket               gcs.Bucket
+	filePath             string
+	fileInfoCache        *lru.Cache
+	sequentialReadSizeMb int32
+	filePerm             os.FileMode
```
Comment (raj-prince) [inline internal/cache/file/downloader/job.go:51]: Should we club all file_creation parameter as a struct? E.g. LocalFileSpec { filePath, filePerm, fileUid, fileGid}

File: internal/cache/file/downloader/job.go [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+func (job *Job) Cancel() {
+	job.mu.Lock()
+	defer job.mu.Unlock()
+}
+
+// Download downloads object till the given offset if not already downloaded
+// and waits for download if waitForDownload is true.
+//
+// Acquires and releases LOCK(job.mu)
+func (job *Job) Download(ctx context.Context, offset int64, waitForDownload bool) (jobStatus JobStatus) {
```
Comment (raj-prince) [inline internal/cache/file/downloader/job.go:137]: Just write a comment TODO, if you want to implement these method later.

File: internal/cache/file/downloader/job.go [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+	job.init()
+	return
+}
+
+// CheckInvariants panic if any internal invariants have been violated.
+func (job *Job) checkInvariants() {
+	// INVARIANT: Each subscriber is of type jobSubscriber
+	for e := job.subscribers.Front(); e != nil; e = e.Next() {
+		switch e.Value.(type) {
+		case jobSubscriber:
```
Comment (raj-prince) [inline internal/cache/file/downloader/job.go:109]: Why switch case? - We will not support types other than jobSubscriber?

**Review by raj-prince — CHANGES_REQUESTED** [review]
Few comments.

File: internal/cache/file/downloader/job.go (hunk 3 above)
Comment (sethiay) [reply internal/cache/file/downloader/job.go:109]: I couldn't find a simple/efficient way to do this using if statement. So, will continue with switch.

Comment (raj-prince) [reply internal/cache/file/downloader/job.go:102]: Tried this - https://go.dev/play/p/0QP63lALIt4

**Review by raj-prince — APPROVED** [review]

**Pull request merged**